src/batMap.ml

```diff
@@ -171,6 +171,99 @@ module Concrete = struct
       | Empty -> raise Not_found in
     loop map
 
+
+  let rec find_first f m =
+    let rec loop_found k0 v0 f = function
+      | Empty -> (k0, v0)
+      | Node (l, k, v, r, _) ->
+         if f k
+         then loop_found k v f l
+         else loop_found k0 v0 f r in
+    match m with
+    | Empty -> raise Not_found
+    | Node (l, k, v, r, _) ->
+       if f k
+       then loop_found k v f l
+       else find_first f r
+
+  let rec find_first_opt f m =
+    let rec loop_found k0 v0 f = function
+      | Empty -> Some (k0, v0)
+      | Node (l, k, v, r, _) ->
+         if f k
+         then loop_found k v f l
+         else loop_found k0 v0 f r in
```

Comment:
Unless I am missing something, the `loop_found` is identical to `loop_found` above. They could be shared by being declared at the toplevel as `find_first_before f k v m`. Same thing with the `loop_found` of the `find_last_*` functions, which could be `find_last_after f k v m`.

Reply:
good finding, thanks. I will refactor this...